molgenis-navigator/src/main/java/org/molgenis/navigator/copy/service/EntityTypeCopier.java

```diff
@@ -0,0 +1,245 @@
+  }
+
+  private EntityType cutDefaultValues(EntityType copy, CopyState state) {
+    stream(copy.getAtomicAttributes())
+        .filter(EntityTypeUtils::isReferenceType)
+        .filter(attr -> attr.getDefaultValue() != null)
+        .forEach(
+            attr -> {
+              state.referenceDefaultValues().put(attr.getIdentifier(), attr.getDefaultValue());
+              attr.setDefaultValue(null);
+            });
+    return copy;
+  }
+
+  private EntityType pasteDefaultValues(EntityType copy, CopyState state) {
+    stream(copy.getAtomicAttributes())
+        .filter(EntityTypeUtils::isReferenceType)
+        .forEach(
+            attr ->
+                attr.setDefaultValue(
+                    state.referenceDefaultValues().getOrDefault(attr.getIdentifier(), null)));
+
+    dataService.getMeta().updateEntityType(copy);
+    return copy;
+  }
+
+  private EntityType persistEntityType(EntityType copy) {
+    dataService.getMeta().addEntityType(copy);
+    return copy;
+  }
+
+  private EntityType copyEntities(EntityType copy, CopyState state) {
+    String originalEntityTypeId = state.originalEntityTypeIds().get(copy.getId());
+    if (!copy.isAbstract()) {
+      dataService
+          .getRepository(originalEntityTypeId)
+          .forEachBatched(batch -> addEntityBatch(copy, state, batch), BATCH_SIZE);
+    }
+    return copy;
+  }
+
+  private void addEntityBatch(EntityType copy, CopyState state, List<Entity> batch) {
+    dataService.add(
+        copy.getId(),
+        batch.stream().map(entity -> new PretendingEntity(entity, state.copiedEntityTypes())));
+  }
+
+  /**
+   * When copying rows from one repository to another, the metadata of these entities will not fit
+   * the copied repository and its references because the metadatas will have different IDs. The
+   * PretendingEntity acts like it's the newly copied entity by returning the metadata of the copied
+   * repository instead of the original.
+   */
+  private class PretendingEntity extends AbstractEntityDecorator {
+
+    private final Map<String, EntityType> copiedEntityTypes;
+
+    PretendingEntity(Entity entity, Map<String, EntityType> copiedEntityTypes) {
+      super(entity);
+      this.copiedEntityTypes = copiedEntityTypes;
+    }
+
+    @Override
+    public EntityType getEntityType() {
+      if (delegate().getEntityType() != null) {
+        String id = delegate().getEntityType().getId();
+        if (copiedEntityTypes.containsKey(id)) {
+          return copiedEntityTypes.get(id);
+        } else {
+          return delegate().getEntityType();
+        }
+      } else {
+        return null;
+      }
+    }
+
+    @Override
+    public Entity getEntity(String attributeName) {
+      Entity entity = delegate().getEntity(attributeName);
+      return entity != null ? newPretendingEntity(entity) : null;
+    }
+
+    /**
+     * Because the File datatype has a reference to {@link FileMetaMetaData} it can happen that a
```

Comment:
Why can't this happen for other types?

Reply:
Within the context of a copy action you never really know which `EntityType`s you're dealing with, so all calls will go through the non-typed getters. That's my theory at least, backed up by the fact that copying the `sys` package (which are all the `EntityType`s that could potentially be requested with a typed getter) does not give problems (all calls go through the non-typed getter).